File: solutions/elixir/rna-transcription/1/rna_transcription.exs
defmodule RNATranscription do
  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RNATranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    do_rna(dna, '')
  end

  defp do_rna('', acc), do: acc
  defp do_rna('G' ++ tail, acc), do: do_rna(tail, acc ++ 'C')
  defp do_rna('C' ++ tail, acc), do: do_rna(tail, acc ++ 'G')
  defp do_rna('T' ++ tail, acc), do: do_rna(tail, acc ++ 'A')
  defp do_rna('A' ++ tail, acc), do: do_rna(tail, acc ++ 'U')
end

File: solutions/erlang/rna-transcription/1/rna_transcription.erl
-module(rna_transcription).
-export([to_rna/1]).

-spec to_rna(string()) -> string().
to_rna(Sequence) ->
  transcribe(Sequence, "").

-spec transcribe(string(), string()) -> string().
transcribe("", Transcription) ->
  Transcription;
transcribe("G" ++ Rest, Transcription) ->
  transcribe(Rest, Transcription ++ "C");
transcribe("C" ++ Rest, Transcription) ->
  transcribe(Rest, Transcription ++ "G");
transcribe("T" ++ Rest, Transcription) ->
  transcribe(Rest, Transcription ++ "A");
transcribe("A" ++ Rest, Transcription) ->
  transcribe(Rest, Transcription ++ "U").